### PR TITLE
Speed up `QGTOnTheFly` (take 2)

### PR DESCRIPTION
**Pull request by attila-i-szabo**
This PR builds on the discussion in #781 to make the logic of `QGTOnTheFly` more performant.
* All the heavy lifting in class `QGTOnTheFlyT` is done by the `Callable` member `mat_vec` which performs the matrix-vector product for pytrees
* This function is generated by the jitted function `mat_vec_factory` as `jax.Partial(mat_vec, jvp_fn)`, where `jvp_fn` is the output of `jax.linearize`, a pytree encoding a linear transformation. This allows us to perform the linearisation work only once per SR step, and optimise the way it's done once and for all.
* In the logic file I chose to separate `mat_vec` from the factory both for aesthetic reasons and because jitting the `Partial` in `solve` and `@` might be more efficient if it's always the same function, not constructed each time inside `mat_vec_factory`

I also fixed the deprecation of `centered`; if it's passed to the upstream version, it breaks the constructor of `QGTOnTheFlyT`

Outstanding issues:
* I've commented out all the tests relevant for `QGTOnTheFly` from `test/optimizer/test_qgt_logic.py` as the logic has completely changed, and I'm not sure what new tests would be appropriate for it.
* I'll implement benchmarks to compare the speed of this implementation to the upstream one. My cluster is down for the weekend, so I'll report back on this later.
* When I can run the benchmarks, I want to check whether jitting `mat_vec` in the logic file would improve performance (I suspect not)

### CODE REVIEW

**Comment by PhilipVinc** [comment]
We went from a pr adding 300 lines of code to one removing 50.

I love that!

(Thanks. I'll look at it soon. I'm also curious on benchmarks.)

File: netket/optimizer/qgt/qgt_onthefly.py [hunk 1]
```diff
@@ -72,26 +83,13 @@ class QGTOnTheFlyT(LinearOperator):
     the field `sr`.
     """
 
-    apply_fun: Callable[[PyTree, jnp.ndarray], jnp.ndarray] = struct.field(
-        pytree_node=False, default=Uninitialized
-    )
-    """The forward pass of the Ansatz."""
+    mat_vec: Callable[[PyTree, float], PyTree] = Uninitialized
+    """The S matrix-vector product as generated by mat_vec_factory.
+    It's a jax.Partial, so can be used as pytree_node."""
```
Comment (PhilipVinc) [inline netket/optimizer/qgt/qgt_onthefly.py:88]: Now that i think of it, all those fields would better be inernal fields... (_mat_vec, _params)

Comment (PhilipVinc) [reply netket/optimizer/qgt/qgt_onthefly.py:88]: because they are not part of the public api (for example, you just removed several)/

**Review by PhilipVinc — COMMENTED** [review]
It looks very good! 

I think the last 2 tests in the file (which test `mat_vec`) can be converted to work with the new code and it would be good to have them.

The other tests, which test fucntions that have been removed, can be safely deleted.

**Comment by PhilipVinc** [comment]
@attila-i-szabo dis you try to benchmark this? I guess it gives similar results to your take 1 implementation?

**Comment by attila-i-szabo** [comment]
Yep, here are the results: https://gist.github.com/attila-i-szabo/2216fcf24cdb76e11a548bb481cb92b1

It is pretty similar to take 1, 20-30% improvement on a GPU (I don't have time to run it on a CPU too, those take forever). However, @chrisrothUT says he finds a 3-4-fold speedup in practice (on GCNNs). I've no clue why it doesn't appear in my tests, maybe it has to do with the ill-conditioned SR problems that you get in a genuine VMC setup. These times are way better than what I've ever encountered in practice, so I suspect something like this is the issue...

**Review by PhilipVinc — APPROVED** [review]
Looks good, thanks!

as for the benchmarks, I just wanted to check that this rewrite did not have some pathological problem.
Seems very good, so that's fine by me.

PS: as for the unexpected speedups, probably you are already doing it but you should try comparing with the same jaxlib version.
Sometimes they fix some operations, sometimes they break some eg: https://github.com/google/jax/issues/7640)

**Comment by PhilipVinc** [comment]
Just wait before merging that I want to check that this does not incur in the same bug as #867 . 
I'll make a pr and add a test asap

**Comment by chrisrothUT** [comment]
> I've no clue why it doesn't appear in my tests, maybe it has to do with the ill-conditioned SR problems that you get in a genuine VMC setup

I didn't run numerical tests, but I just remember SR being a deadly bottleneck on occasion whereas now it seems to always take on the order of sampling and computing expectation values. I do think that the difference is probably more pronounced on ill-conditioned problems especially if `time_old/time_new` scales something like `(startup + runtime2)/ (startup + runtime) `and startup is significant for these random matrices.

**Comment by attila-i-szabo** [comment]
> Just wait before merging that I want to check that this does not incur in the same bug as #867 .
> I'll make a pr and add a test asap

In some sense it does because JVP/VJP cut away any imaginary part in the gradients associated with real parameters. It doesn't error out but it drops the imag. parts silently. You need this behaviour to get S = Re(ΔO^H ΔO) so we couldn't dispense with it even if autodiff somehow allowed it.

**Comment by PhilipVinc** [comment]
>In some sense it does because JVP/VJP cut away any imaginary part in the gradients associated with real parameters. It doesn't error out but it drops the imag. parts silently. You need this behaviour to get S = Re(ΔO^H ΔO) for real parameters so we couldn't dispense with it even if autodiff somehow allowed it.

If you use `nk.jax.vjp` this does not happen (and the imaginary part is computed).

Is this also valid for `jvp`?

**Comment by PhilipVinc** [comment]
For posterity: to make this work with complex vectors in the matmul we need to 

> by the way, the new implementation does not use jax.vjp anymore, relying instead on jax.linearize ... so i guess the solution here would be to write nk.jax.linearize to work somehow like nk,jax.vjp and not drop the imaginary part, and use that inside of QGTOnbTheFly

**Comment by attila-i-szabo** [comment]
cool. I guess this is mergeable now?

**Comment by attila-i-szabo** [comment]
Tests are added back now. Do I need to rebase, or can it be merged as it is?

**Comment by PhilipVinc** [comment]
should be good.
if tests pass I'll merge
